libs/agno/agno/models/ibm/watsonx.py

```
@@ -0,0 +1,382 @@
+    """
+    Format a message into the format expected by WatsonX.
+
+    Args:
+        message (Message): The message to format.
+
+    Returns:
+        Dict[str, Any]: The formatted message.
+    """
+    if message.images is not None and not isinstance(message.content, str):
+        message = add_images_to_message(message=message, images=message.images)
+    return message.serialize_for_model()
+
+
+@dataclass
+class WatsonX(Model):
+    """
+    A class for interacting with IBM WatsonX models.
+
+    See supported models at: https://dataplatform.cloud.ibm.com/docs/content/wsj/analyze-data/fm-models.html?context=wx
+
+    For more information, see: https://www.ibm.com/watsonx/developer/
+    """
+
+    id: str = "ibm/granite-20b-code-instruct"
+    name: str = "WatsonX"
+    provider: str = "IBM"
+
+    # Request parameters
+    frequency_penalty: Optional[float] = None
+    presence_penalty: Optional[float] = None
+    max_tokens: Optional[int] = None
+    temperature: Optional[float] = None
+    top_p: Optional[float] = None
+    logprobs: Optional[int] = None
+    top_logprobs: Optional[int] = None
+    response_format: Optional[Any] = None
+
+    request_params: Optional[Dict[str, Any]] = None
+
+    # Client parameters
+    api_key: Optional[str] = None
+    project_id: Optional[str] = None
+    url: Optional[str] = "https://eu-de.ml.cloud.ibm.com"
+    verify: bool = True
+    client_params: Optional[Dict[str, Any]] = None
+
+    # WatsonX client
+    model_client: Optional[ModelInference] = None
+
+    def _get_client_params(self) -> Dict[str, Any]:
+        # Fetch API key and project ID from env if not already set
+        self.api_key = self.api_key or getenv("IBM_WATSONX_API_KEY")
+        if not self.api_key:
+            logger.error("IBM_WATSONX_API_KEY not set. Please set the IBM_WATSONX_API_KEY environment variable.")
+            raise ValueError("IBM_WATSONX_API_KEY environment variable not set")
```

We usually let the client throw the API key not found error. What is the best practice here?

---

Yeah good point, probably best if we raise it? But for now lets just be consistent and improve all later